[InstCombine] Fold max(max(x, c1) << c2, c3) —> max(x << c2, c3) when c3 >= c1 * 2 ^ c2

Charukesh827 opened the pull request:
As suggested generalize to fold max(max(x, c1) binop c2, c3) —> max(x binop c2, c3) if c3>=C1* 2 ^ c2 is done.

define i8 @src(i8 %arg0) {
  %1 = call i8 @llvm.umax.i8(i8 %arg0, i8 1)
  %2 = shl nuw i8 %1, 1
  %3 = call i8 @llvm.umax.i8(i8 %2, i8 16)
  ret i8 %3
}

define i8 @tgt(i8 %arg0) {
  %1 = shl nuw i8 %arg0, 1
  %2 = call i8 @llvm.umax.i8(i8 %1, i8 16)
  ret i8 %2
}

Closes https://github.com/llvm/llvm-project/issues/139786.

Charukesh827 committed:
Add test for shifting binop  in InstCombine transformation for the issue  Missed Optimization: max(max(x, c1) << c2, c3) —> max(x << c2, c3) when c3 >= c1 * 2 ^ c2 #139786

Charukesh827 committed:
[InstCombine] Fix Missed Optimization: max(max(x, c1) << c2, c3) —> max(x << c2, c3) when c3 >= c1 * 2 ^ c2

This patch fixes issue #139786
 where InstCombine where it Missed Optimization: max(max(x, c1) << c2, c3) —> max(x << c2, c3) when c3 >= c1 * 2 ^ c2. Pre-committed test in <commit-hash>.

Alive2: https://alive2.llvm.org/ce/z/on2tJE

Charukesh827 committed:
removed unwanted header (iostream)

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:
```diff
@@ -1171,6 +1171,84 @@ static Instruction *moveAddAfterMinMax(IntrinsicInst *II,
+     return nullptr;
+
+  // Check if BinOp is a left shift
+  if (BinOp != Instruction::Shl) {
```
IIRC, you are trying to implement solution 2 suggested by me: https://github.com/llvm/llvm-project/issues/139786#issuecomment-2878842283

If it is the case, you should generalize it to handle most of binops (excluding div/rem), then use `simplifyBinOp` and `simplifyBinaryIntrinsic` to check if `min/max(c1 binop c2, c3)` folds to `c3`.

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:
```diff
@@ -1171,6 +1171,84 @@ static Instruction *moveAddAfterMinMax(IntrinsicInst *II,
+  auto* BinOpInst = cast<BinaryOperator>(Op0);
+  Instruction::BinaryOps BinOp = BinOpInst->getOpcode();
+  Value *X;
+  InnerMax = BinOpInst->getOperand(0);
+  // std::cout<< InnerMax->dump() <<std::endl;
+  if(!match(InnerMax,m_OneUse(m_Intrinsic<Intrinsic::umax>(m_Value(X),m_APInt(C))))){
+  if(!match(InnerMax,m_OneUse(m_Intrinsic<Intrinsic::smax>(m_Value(X),m_APInt(C))))){
+  if(!match(InnerMax,m_OneUse(m_Intrinsic<Intrinsic::umin>(m_Value(X),m_APInt(C))))){
+  if(!match(InnerMax,m_OneUse(m_Intrinsic<Intrinsic::smin>(m_Value(X),m_APInt(C))))){
+     return nullptr;
+  }}}}
+  
+  auto *InnerMaxInst = cast<IntrinsicInst>(InnerMax);
+
+  bool IsSigned = MinMaxID == Intrinsic::smax || MinMaxID == Intrinsic::smin;
```
```suggestion
  auto *InnerMinMaxInst = dyn_cast<MinMaxIntrinsic>(BinOpInst->getOperand(0));
  if (!InnerMinMaxInst || !InnerMinMaxInst->hasOneUse()) return nullptr;

  bool IsSigned = InnerMinMaxInst->isSigned();
```

dtcxzyw commented on llvm/test/Transforms/InstCombine/shift-binop.ll:
```diff
@@ -0,0 +1,27 @@
+  ret i8 %3
+}
+
+define i8 @tgt(i8 %arg0) {
```
`tgt` is not necessary. It has been checked by the check lines for `@src`.

dtcxzyw commented on llvm/test/Transforms/InstCombine/shift-binop.ll:
```diff
@@ -0,0 +1,27 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py
```
Can you please add more negative tests/vector tests/multi-use tests, as suggested by https://llvm.org/docs/InstCombineContributorGuide.html#tests?

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:
```diff
@@ -1171,6 +1171,84 @@ static Instruction *moveAddAfterMinMax(IntrinsicInst *II,
+  
+  //Create new min/max intrinsic with new binop and c3
+  
+    if(IsSigned){
```
Please run `clang-format` before committing your code.

dtcxzyw replied:
```suggestion
    if (auto *NewBinInst = dyn_cast<Instruction>(NewBinOp))
    if(IsSigned){
```
`NewBinOp` may be constant-folded.

dtcxzyw reviewed (changes requested):
Please add the alive2 proof into the PR description.

dtcxzyw commented on llvm/test/Transforms/InstCombine/shift-binop.ll:
```diff
@@ -0,0 +1,27 @@
+  ret i8 %2
+}
+
+declare i8 @llvm.umax.i8(i8, i8)
```
```suggestion
```
Intrinsic declarations are unnecessary.

dtcxzyw commented on llvm/test/Transforms/InstCombine/shift-binop.ll:
```diff
@@ -0,0 +1,27 @@
+  %2 = call i8 @llvm.umax.i8(i8 %1, i8 16)
+  ret i8 %2
+}
+
```
Add the original motivating case from https://github.com/llvm/llvm-project/issues/139786#issue-3061114504.

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:
```diff
@@ -1171,6 +1171,84 @@ static Instruction *moveAddAfterMinMax(IntrinsicInst *II,
+  if((IsSigned && !BinOpInst->hasNoSignedWrap()) ||
+     (!IsSigned && !BinOpInst->hasNoUnsignedWrap())) 
+     return nullptr;
+
```
Should check that inner min/max has the same intrinsic ID as the outer one.

Charukesh827 committed:
added negative test

Charukesh827 committed:
Made suggested changes but couldn't generalize to div/rem as suggested

"If it is the case, you should generalize it to handle most of binops (excluding div/rem), then use simplifyBinOp and simplifyBinaryIntrinsic to check if min/max(c1 binop c2, c3) folds to c3."

Charukesh827 committed:
added motivation to shift-binop.ll

Charukesh827 commented:
made most of suggested changes. Only thing is i didn't understand what has to be done for generalizing div, Please help me with that. 

about the suggestion you gave:

"Solution1: Canonicalize max(x << shamt, c << shamt) into max(x, c) << shamt: https://alive2.llvm.org/ce/z/mQEDAQ
We already did similar things in moveAddAfterMinMax .

Solution2: Generalize to fold max(max(x, c1) binop c2, c3) —> max(x binop c2, c3)."

solution 1 was already available so i didn't make it. I only concentrated on solution 2

![image](https://github.com/user-attachments/assets/f0760e29-f8ce-4549-82b4-d8a2a2daeac7)

Alive2: https://alive2.llvm.org/ce/z/on2tJE

Charukesh827 committed:
Merge branch 'main' into main

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:
```diff
@@ -1174,6 +1174,86 @@ static Instruction *moveAddAfterMinMax(IntrinsicInst *II,
                   : BinaryOperator::CreateNUWAdd(NewMinMax, Add->getOperand(1));
 }
+
+// Try canonicalize max(max(X,C1) binop C2, C3) -> max(X binop C2, C3)
```
```suggestion
/// Try canonicalize max(max(X,C1) binop C2, C3) -> max(X binop C2, C4)
```
Use `///` for header comments.

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:
```diff
@@ -1174,6 +1174,86 @@ static Instruction *moveAddAfterMinMax(IntrinsicInst *II,
 }
+
+// Try canonicalize max(max(X,C1) binop C2, C3) -> max(X binop C2, C3)
+static Instruction *moveShiftAfterMinMax(IntrinsicInst *II,
```
The function name should be updated.
In fact, this fold can be decomposed into two steps:
```
max(max(X,C1) binop C2, C3) -> // Associative laws
max(max(X binop C2, C1 binop C2), C3) -> // Commutative laws
max(X binop C2, max(C1 binop C2, C3)) -> // Constant fold
max(X binop C2, C4)
```

dtcxzyw replied:
`max(X, C1) binop C2 -> max(X binop C2, C1 binop C2)` is not always safe for all binops. You can reuse the helper `leftDistributesOverRight`.

dtcxzyw commented on llvm/test/Transforms/InstCombine/shift-binop.ll:
```diff
@@ -0,0 +1,42 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py
+; RUN: opt < %s -passes=instcombine -S | FileCheck %s
+
+;define i32 @src(i32 %x, i32 %shamt, i32 %c) {
+;  %shl = shl nuw i32 %x, %shamt
+;  %c2 = shl nuw i32 %c, %shamt
+;  %max = call i32 @llvm.umin(i32 %shl, i32 %c2)
+;  ret i32 %max
+;}
+
+;define i32 @tgt(i32 %x, i32 %shamt, i32 %c) {
+;  %max = call i32 @llvm.umin(i32 %x, i32 %c)
+;  %shl = shl i32 %max, %shamt
+;  ret i32 %shl
+;}
```
```suggestion
```

dtcxzyw commented:
> solution 1 was already available so i didn't make it.

`c << shamt` has been constant folded so that instcombine doesn't catch this pattern. Godbolt: https://godbolt.org/z/rndGjEE31

> I only concentrated on solution 2

Yeah. I think it is a better solution.

Charukesh827 committed:
Merge branch 'main' into main

Charukesh827 committed:
made the suggested changes as follows:

1)max(X, C1) binop C2 -> max(X binop C2, C1 binop C2) is not always safe for all binops. You can reuse the helper leftDistributesOverRight.
2)The function name should be updated.
In fact, this fold can be decomposed into two steps:

max(max(X,C1) binop C2, C3) -> // Associative laws
max(max(X binop C2, C1 binop C2), C3) -> // Commutative laws
max(X binop C2, max(C1 binop C2, C3)) -> // Constant fold
max(X binop C2, C4)

Charukesh827 committed:
updated the test for the changed opt

Charukesh827 commented:
I wrote rightDistributesOverLeft function with the knowledge i have on this opt, please let me know is anything to be added or removed.

I considered only Add, Sub, Mul, Shl operation, as i was able to verify only these.

another optimization(i didn't write) dominates this optimization
Alive2 : [https://alive2.llvm.org/ce/z/sMAnf_](https://alive2.llvm.org/ce/z/sMAnf_)

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:
```diff
@@ -1174,6 +1174,163 @@ static Instruction *moveAddAfterMinMax(IntrinsicInst *II,
 }
+
+
+static bool rightDistributesOverLeft(Instruction::BinaryOps ROp, bool HasNUW,
```
Please add some header comments for this function.

dtcxzyw replied:
Can you provide alive2 proof for this function (`(X LOp Y) ROp Z -> (X ROp Z) LOp (Y ROp Z)`)?

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:
```diff
@@ -1174,6 +1174,163 @@ static Instruction *moveAddAfterMinMax(IntrinsicInst *II,
+///  Associative laws max(max(X binop C2, C1 binop C2), C3) -> // Commutative
+///  laws max(X binop C2, max(C1 binop C2, C3)) -> // Constant fold max(X binop
+///  C2, C4)
+
```
```suggestion
```

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:
```diff
@@ -1174,6 +1174,163 @@ static Instruction *moveAddAfterMinMax(IntrinsicInst *II,
+  case Intrinsic::smin:
+    // Signed min/max distribute over addition if no signed wrap.
+    if (HasNSW && ROp == Instruction::Add)
+      return true;
```
It doesn't hold for smin/smax: https://alive2.llvm.org/ce/z/XFf_U8

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:
```diff
@@ -1174,6 +1174,163 @@ static Instruction *moveAddAfterMinMax(IntrinsicInst *II,
+
+  Value *Op0 = II->getArgOperand(0), *Op1 = II->getArgOperand(1);
+  Value *InnerMax;
+  const APInt *C;
+  if (!match(Op0, m_OneUse(m_BinOp(m_Value(InnerMax), m_APInt(C)))) ||
+      !match(Op1, m_APInt(C)))
```
```suggestion
  Constant *C2, *C3;
  if (!match(Op0, m_OneUse(m_BinOp(m_Value(InnerMax), m_ImmConstant(C2)))) ||
      !match(Op1, m_ImmConstant(C3)))
```

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:
```diff
@@ -1174,6 +1174,163 @@ static Instruction *moveAddAfterMinMax(IntrinsicInst *II,
+    C1BinOpC2 = IsSigned ? C1.smul_ov(C2, overflow) : C1.umul_ov(C2, overflow);
+    break;
+  case Instruction::Sub:
+    C1BinOpC2 = IsSigned ? C1.ssub_ov(C2, overflow) : C1.usub_ov(C2, overflow);
+    break;
+  case Instruction::Shl:
+    // Compute C1 * 2^C2
+    Two = APInt(C2.getBitWidth(), 2);
+    Pow2C2 = Two.shl(C2);        // 2^C2
+    C1TimesPow2C2 = C1 * Pow2C2; // C1 * 2^C2
+
+    // Check C3 >= C1 * 2^C2
+    if (C3.ult(C1TimesPow2C2)) {
+      return nullptr;
+    } else {
+      C1BinOpC2 = C1.shl(C2);
+    }
+    break;
+  default:
+    return nullptr; // Unsupported binary operation
+  }
+
+  // Constant fold: Compute MinMaxID(C1 binop C2, C3) to get C4
+  APInt C4;
+  switch (MinMaxID) {
+  case Intrinsic::umax:
+    C4 = APIntOps::umax(C1BinOpC2, C3);
+    break;
+  case Intrinsic::umin:
+    C4 = APIntOps::umin(C1BinOpC2, C3);
+    break;
+  case Intrinsic::smax:
+    C4 = APIntOps::smax(C1BinOpC2, C3);
+    break;
+  case Intrinsic::smin:
+    C4 = APIntOps::smin(C1BinOpC2, C3);
+    break;
+  default:
+    return nullptr; // Unsupported intrinsic
+  }
```
```suggestion
    Constant *C1;
    if (!match(InnerMinMaxInst->getRHS(), m_ImmConstant(C1))
      return nullptr;
    Constant *C1BinOpC2 = ConstantFoldBinaryOpOperands(BinOp, C1, C2, DL);
    Constant *C4 = ConstantFoldBinaryIntrinsic(MinMaxID, C1BinOpC2, C3, C3->getType(), nullptr);
```

Charukesh827 commented:
If we use this [Alive2 proof](https://alive2.llvm.org/ce/z/sS_Urz) then our aim "Fold max(max(x, c1) << c2, c3) —> max(x << c2, c3) when c3 >= c1 * 2 ^ c2" itself becomes false, Shouldn't we do the Alive2 experiment with only one x and all other constant(like [Alive2](https://alive2.llvm.org/ce/z/wR826y))?

Charukesh827 committed:
made all suggested changes

Charukesh827 commented:
made all suggested changes, the rightDistributesOverLeft follows this alive [proof](https://alive2.llvm.org/ce/z/25s-3Z)

Charukesh827 committed:
made ashr/lshr generic  bu removing it from nuw condition.
added positive and negative testes
added some multiuse testes
